### PR TITLE
[SPARK-19220][UI] Make redirection to HTTPS apply to all URIs.

### PR DESCRIPTION
**Pull request by vanzin**
The redirect handler was installed only for the root of the server;
any other context ended up being served directly through the HTTP
port. Since every sub page (e.g. application UIs in the history
server) is a separate servlet context, this meant that everything
but the root was accessible via HTTP still.

The change adds separate names to each connector, and binds contexts
to specific connectors so that content is only served through the
HTTPS connector when it's enabled. In that case, the only thing that
binds to the HTTP connector is the redirect handler.

Tested with new unit tests and by checking a live history server.

### CODE REVIEW

**Comment by vanzin** [comment]
Pinging some (random?) people @ajbozarth @srowen @sarutak

**Comment by ajbozarth** [comment]
The code changes looks good to me, but my experience in code working with SSL is still small so someone with more experience should also double-check.

**Comment by sarutak** [comment]
I'll take a look at this within the weekend.

**Comment by vanzin** [comment]
ping?

**Comment by sarutak** [comment]
@vanzin I'm looking into this change and it works well on standalone-mode but doesn't on yarn-mode.
I think it is because ResourceManager's web proxy might not handle https properly.
It seems httpclient in `WebAppProxyServlet` is not configured for SSL.
Do you have any idea?

The change itself seems good to me.

**Comment by vanzin** [comment]
> I think it is because ResourceManager's web proxy might not handle https properly.

Yeah, that's a known issue with enabling SSL for the web UI on YARN with self-signed certificates.

**Comment by sarutak** [comment]
I understand. if there are no additional comments from anyone by tomorrow, I'll merge this.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 1]
```diff
@@ -306,23 +311,31 @@ private[spark] object JettyUtils extends Logging {
...
+          val scheme = "https"
+          // Create a connector on port securePort to listen for HTTPS requests
+          val connector = new ServerConnector(server, factory)
+          connector.setPort(securePort)
+          connector.setName(SPARK_CONNECTOR_NAME)
+          connectors += connector
+
+          // redirect the HTTP requests to HTTPS port
+          httpConnector.setName(REDIRECT_CONNECTOR_NAME)
+          collection.addHandler(createRedirectHttpsHandler(securePort, scheme))
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: I noticed one point.
If a port is already used, `collection.addHandler` will take place more than twice leading redirection doesn't work properly.
Of course, it's not your fault. If you fix it in this PR together, it's good but it's a separate issue so I'll fix in another PR otherwise.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Actually that is caused by my change, so let me fix it.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 2]
```diff
@@ -337,17 +350,20 @@ private[spark] object JettyUtils extends Logging {
...
 
       val errorHandler = new ErrorHandler()
       errorHandler.setShowStacks(true)
       errorHandler.setServer(server)
       server.addBean(errorHandler)
+
+      gzipHandlers.foreach(collection.addHandler)
       server.setHandler(collection)
+
+      server.setConnectors(connectors.toArray)
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Why did you move `server.setConnectors(connectors.toArray)` and `gzipHandlers.foreach(collection.addHandler)`?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala [hunk 3]
```diff
@@ -274,25 +277,28 @@ private[spark] object JettyUtils extends Logging {
       conf: SparkConf,
       serverName: String = ""): ServerInfo = {
 
-    val collection = new ContextHandlerCollection
     addFilters(handlers, conf)
 
     val gzipHandlers = handlers.map { h =>
+      h.setVirtualHosts(Array("@" + SPARK_CONNECTOR_NAME))
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Do we need this code here? `setVirtualHosts` should always be called in `addHandler` for each handler right?

**Review by sarutak — COMMENTED** [review]
It almost LGTM.
I leave some comments and have one question.
Why not simply remove old redirect handler like `collection.removeHandler` ?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 2 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Mostly for grouping. "This is where all handlers are added to the server."

In any case I have another change (#16625) that kinda moves all this stuff around anyway...

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 3 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: `ContextHandlerCollection.addHandler` does not call `setVirtualHosts`.

**Comment by vanzin** [comment]
> Why not simply remove old redirect handler like collection.removeHandler ?

I find it cleaner to just not do something that to do it then have to undo things when they fail.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 3 above)
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: `addHandler` here means `ServerInfo#addHandler`, sorry for confusing.
And I noticed `serverInfo` in `WebUI` can be `None` and `serverInfo.foreach(_.addHandler(handler))` is not called within `WebUI#attachHandler` in that case. So I understand it's reasonable change.
Can you add a test case for justification of this change?

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 2 above)
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: O.K. That's what I thought.

**Comment by sarutak** [comment]
O.K, It's reasonable.

File: core/src/main/scala/org/apache/spark/ui/JettyUtils.scala (hunk 3 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: I don't understand what you mean. Without this change the UI does not work at all. The test I added already covers it.

Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: > Without this change the UI does not work at all. The test I added already covers it.
Hmm, it's funny. I commented out this change and run the test case you added (UISuite and UISeleniumSuite) but it passed.

Comment (sarutak) [inline core/src/main/scala/org/apache/spark/ui/JettyUtils.scala]: Of course I understand this change is needed. I've confirmed it manually.

File: core/src/test/scala/org/apache/spark/ui/UISuite.scala [hunk 4]
```diff
@@ -227,8 +228,55 @@ class UISuite extends SparkFunSuite {
...
+      def newContext(path: String): ServletContextHandler = {
+        val ctx = new ServletContextHandler()
+        ctx.setContextPath(path)
+        ctx.addServlet(new ServletHolder(servlet), "/*")
+        ctx
+      }
+
+      val (conf, sslOptions) = sslEnabledConf()
+      serverInfo = JettyUtils.startJettyServer(
+        "0.0.0.0", 0, sslOptions, Seq[ServletContextHandler](newContext("/")), conf)
```
Comment (sarutak) [inline core/src/test/scala/org/apache/spark/ui/UISuite.scala]: To test the neseccity of `setVirtualHost` in `JettyUtils#StartJettyServer` correctly, you might need to add another ServletContextHandler instance like `newContext("/test0")` and corresponding assertion.

Comment (vanzin) [inline core/src/test/scala/org/apache/spark/ui/UISuite.scala]: See commit 67df755 (the second commit in this PR). Tests failed before that commit.

**Comment by sarutak** [comment]
The latest change LGTM. Merging into `master`. Thanks @vanzin !

**Comment by sarutak** [comment]
This change cannot be applied to `branch-2.0` and `branch-2.1` cleanly so please open other PRs for those branches. Thanks.

**Comment by vanzin** [comment]
I wasn't really planning to backport this, unless someone is really interested in this functionality.
